File: solutions/c/saddle-points/1/saddle_points.c
#include "saddle_points.h"
#include <stdio.h>

saddle_points_t* saddle_points(uint8_t rows, uint8_t columns, uint8_t matrix[rows][columns]){
  
    saddle_points_t* sps = malloc(sizeof(saddle_points_t));
    if (matrix==NULL){
        sps->count=0;
        return sps;
    }    
    sps->points = malloc(0);
    size_t spi=-1;
    uint8_t rowmax;
    uint8_t maxcol;
    uint8_t maxrow;
    for (uint8_t r=0; r<rows; r++){
        rowmax=0;
        for (uint8_t c=0; c<columns; c++) {
            //find max value in row
            if (matrix[r][c]>rowmax) {
                rowmax=matrix[r][c];
                maxcol=c;
            }
        }
        //check if max val in row is min val in col
        uint8_t dr;
        for (dr=0; dr<rows; dr++) {
            if (dr!=r && matrix[dr][maxcol]<rowmax) break;
        }
        if (dr==rows) {
            maxrow=r;
            break; 
        }  else {
            rowmax=0;
        }  
    }
    
    if (rowmax>0) {                
        spi++;
        sps->points = realloc(sps->points,(spi+1)*sizeof(saddle_point_t));
        sps->count++;
        sps->points[spi].row=maxrow+1;
        sps->points[spi].column=maxcol+1;

        for (uint8_t i=maxcol+1;i<columns;i++) {
            if (matrix[maxrow][i]==rowmax) {                
                spi++;
                sps->points = realloc(sps->points,(spi+1)*sizeof(saddle_point_t));
                sps->count++;
                sps->points[spi].row=maxrow+1;
                sps->points[spi].column=i+1;
            }
        }
        if (spi<1) {
            for (uint8_t i=maxrow+1;i<rows;i++) {
                if (matrix[i][maxcol]==rowmax) {                
                    spi++;
                    sps->points = realloc(sps->points,(spi+1)*sizeof(saddle_point_t));
                    sps->count++;
                    sps->points[spi].row=i+1;
                    sps->points[spi].column=maxcol+1;
                }
            }           
        } 
    }
    return sps;

}

void free_saddle_points(saddle_points_t* sp){
    free(sp);
}

File: solutions/fortran/saddle-points/1/saddle_points.f90
module saddle_points
  implicit none

  type :: point_t
    integer :: row
    integer :: column
  end type point_t

contains

  function saddlePoints(matrix) result(points)
    integer, intent(in) :: matrix(:, :)
    type(point_t), allocatable :: points(:)
    type(point_t), dimension(10) :: tpoints
    integer :: i, j, k, m, mr, mi, np
    !print *,"=========== array ========="
    !do i =1,size(matrix,1)
    !  print *, matrix(i,:)
    !end do  
    np = 0 
    ! For all teh rows
    do i = 1,size(matrix,1)
       ! Find the largest value in a row
       mr = matrix(i,1)
       mi = 1
       do j = 2,size(matrix,2)
          if (matrix(i,j) > mr) then
             mi = j
             mr = matrix(i,j)
          end if   
       end do
       !print *, "mr=", mr
       ! Check if max row is min in column
       do k = 1, size(matrix, 1)
          !print *, "mi=",mi,"k=",k,",matrix(k,mi)=",matrix(k,mi)
          if (matrix(k,mi) < mr) exit
       end do
       !print *, "k=", k   
       if (k > size(matrix, 1)) then
           np = np + 1    
           !print *, "r=",i,",c=",mi  
           tpoints(np) = point_t(i , mi)
       end if
       ! Check if there are other columns with the same max row value
       !print *,"check the other cols if equal"
       do k=mi+1, size(matrix, 2)
          !print *, "second check, k=", k, ",matrix(i,k)=",matrix(i,k)
          if (matrix(i,k) == mr) then
             !check if min in column     
              do m = 1, size(matrix, 1)
                if (matrix(m,k)<mr) exit
              end do  
              if (m > size(matrix, 1)) then
                np  = np + 1    
                !print *, "r=",i,",c=",k  
                tpoints(np) = point_t(i , k)
              end if
          end if        
       end do   
    end do    
    allocate(points(np))
    points(1:np) = tpoints(1:np)
  end function saddlePoints

end module saddle_points
